### ISSUE TITLE
Enhancement to the 'Tree' object

### ISSUE THREAD

**Issue by MarqueIV** [umlet/umlet#512]
Hi! Mark again. I'm the one trying to help you figure out the issues on Mac with BetterTouchTool, etc.  Anyway, tonight while using your app, I came across your 'Tree' element, where you define the visual tree via indenting with tabs.  I see you have 'type=Package' and it shows folders.  Remove that line and it shows stick-figures.

Well I have an idea that I'd be surprised if you haven't already considered considering your properties window essentially uses it for everything else: when defining a tree structure, using prefixes to determine the node type/graphic!

I whipped up something in C#/WPF tonight to see the viability of what I was talking about (and becauseI thought it would be fun to try and write a tab-based tree parser.) Take a look at my attached screenshot.  In my case, I just used a small 'Type bubble' because I was more testing that parsing than the rendering, but it proves the point quite nicely.  

I have to say, I love the way it came out and would *love* to see you integrate something similar into UMLet.  I love how it lets me 'rough out' my object hierarchies.

In my example, the prefix types I use are as follows:

(No prefix) = Object (Your standard fare object.)
'-' = Property (straight line because it's straightforward)
'~' = Method (wavy because the line is 'doing something', just like a method)
'#' = Collection ('Number' because there's a number of them)

Those are just what I came up with as a proof-of-concept. No idea if I'm misusing them, but you get the idea.

Anyway, I'd love to know what you think!  Lemme know and let me know if you want me to share my parsing code!

<img width="854" alt="nodetypesviaprefixes" src="https://user-images.githubusercontent.com/897883/44765989-4edc4c00-ab25-11e8-8456-ffa96f388c7d.png">

**Comment by afdia** [comment]
If I understand it correctly, you want to use a prefix (- ~ #) which is transformed to a marker which specifies the type of the node?
Or let's say it differently you want to write `-Name` instead of e.g. `Name [Property]` and perhaps get a more beautiful marker (because if you write it yourself it's the same color and size as the "Name")
It would probably only make sense for type=Package but not for Actor or WorkProcess.

It could be an interesting feature, but it's another special syntax the user have to learn and it's not very flexible (e.g. what if I want to specify the size of the collection or additional information about methods, properties, ...)

Intuitively I would just use a syntax similar to [UML Class diagrams](https://en.wikipedia.org/wiki/Class_diagram) to mark something as methods, properties or collections. E.g.:
Property (you could also specify the type after a colon like in UML)
Method()
Collection[] (you could also use UML Class style syntax to specify an amount: [1..*]

I'm not sure why you would separate properties and objects, but everything else could be expressed already (and optionally with more details) using UML Class like syntax.
e.g. a simple example:
![grafik](https://user-images.githubusercontent.com/8770707/44770051-02196500-ab67-11e8-8fcb-3153fa5c4b30.png)

What do you think?

Btw: Umlet parses the code and triggers the handleLine method in https://github.com/umlet/umlet/blob/master/umlet-elements/src/main/java/com/baselet/element/facet/specific/HierarchyFacet.java for every line (there you see the 3 if/else statements which draw the different symbols

**Comment by MarqueIV** [comment]
Close.  More simple though.  I'm more only talking about what icon is rendered to the left of the text, nothing more. (My bad for putting the labels to the right instead of the left as I can see how that confuses my intent.)

I'm actually envisioning some sort of 'prefix-to-png/svg' list that you can set up in the software, maybe just for that tree, but perhaps something that can be incorporated in any control--so when you type a node that starts with it, instead of the folder icon, you get whatever png/svg is mapped to that prefix.  

Then, by using the agreed-upon prefixes for UML diagrams, you can already handle those graphics internally.  I don't know them off the top of my head, but let's say an interface is represented by the '=' prefix.  If I typed '=IMyInterface' it would render with a circle for the node (instead of a folder), then 'IMyInterface'.  Make sense?

Point being you could set up defaults that match the UML standard, but with a simple prefix-to-filename list somewhere, you can let a user easily render any graphic at any node in the tree, just by using a prefix from that list.  Make sense?

If not, I can whip up an example in Illustrator to show what I'm talking about.

**Comment by afdia** [comment]
I see, so this is more about getting the flexibility to render any image as part of a diagram element?
Thats an interesting idea, the only disadvantage is that now you have to share the diagram and the used images (so sharing the uxf alone is no longer enough to make the diagram work for everyone), but if it would support URLs, it would still work as long as those URLs can be resolved from the given machine.

You should probably rename the image to "add possibility to include images (svg/png/...) into a diagram element" or something like that.

One difficulty when implementing this would be how to handle overlapping images? Either we would have to move the next textline by x pixels to make sure the image fits or we could specify that overlapping is ok and the user has to move the next textline by adding some linebreaks (in case the image is higher than one line)

**Comment by MarqueIV** [comment]
Yes, it's just about having flexibility with the images.  You're correct it isn't as nice if the files are external, but if a person is putting in a file reference, I'm sure they know that.

Another approach, in addition to the above, would be to add a bunch of pre-defined images that the user can choose from.  All the standard ones that you're already using as part of other controls (i.e. a folder, stick figure, add others like circles, squares, stacked, offset squares, 'X'es, etc.). For even more flexibility, let a person reference another element in the diagram.  That way they define the elements in one part, then reference them in another. (You could even move these to a separate pallet so they don't appear in the actual diagram.)

Anyway, I'm getting away from myself with all the potential ways this could go.  The TLDR version is, and the entire reason for this is that I'd like to choose--at least from a minimum, pre-defined set--which graphic appears at each node in the tree.  I'll update the title as requested.

**Comment by MarqueIV** [comment]
Haven't forgotten about you guys there.  I will get to this. In the meantime, I've posted a new suggestion for your too.  LMK what you think!
